Core/include/Acts/Geometry/DetectorVolume.hpp

```diff
@@ -0,0 +1,577 @@
+  return m_volumeMaterial;
+}
+
+inline const IVolumeMaterial* DetectorVolume::volumeMaterial() const {
+  return m_volumeMaterial.get();
+}
+
+inline const GeometryIdentifier& DetectorVolume::geometryId() const {
+  return m_geometryId;
+}
+
+inline const std::string& DetectorVolume::name() const {
+  return m_name;
+}
+
+inline void DetectorVolume::assignDetector(const Detector& detector) {
+  m_detector = &detector;
+}
+
+inline const Detector* DetectorVolume::detector() const {
+  return m_detector;
+}
+
+/// @brief  A detector volume factory which first constructs the detector volume
+/// and then constructs the portals. This ensures that the std::shared_ptr
+/// holding the detector volume is not weak when assigning to the portals.
+class DetectorVolumeFactory {
+ public:
+  /// Create a detector volume - from factory
+  /// @param portalGenerator the volume portal generator
+  /// @param gctx the geometry context for construction and potential contextual store
+  /// @param args the arguments forwarded to the detector volume contructor
+  ///
+  /// @return a shared object DetectorVolume
+  template <typename... Args>
+  static std::shared_ptr<DetectorVolume> construct(
+      const PortalGenerator& portalGenerator, const GeometryContext& gctx,
+      Args&&... args) {
+    auto dVolume =
+        DetectorVolume::makeShared(gctx, std::forward<Args>(args)...);
+    dVolume->construct(gctx, portalGenerator);
+    return dVolume;
+  }
+};
+
+/// Helper extractors: all portals
+struct AllPortalsExtractor {
+  /// Extract the portals from the volume
+  ///
+  /// @param gctx the geometry contextfor this extraction call
+  /// @param nState is the current navigation state
+  ///
+  /// @return a vector of raw Portal pointers
+  inline static const std::vector<const Portal*> extract(
+      [[maybe_unused]] const GeometryContext& gctx,
+      const NavigationState& nState) {
+    if (nState.currentVolume == nullptr) {
+      throw std::runtime_error(
+          "AllPortalsExtractor: no detector volume given.");
+    }
+    return nState.currentVolume->portals();
+  }
+};
+
+/// Helper extractors: all surfaces
+struct AllSurfacesExtractor {
+  /// Extract the surfaces from the volume
+  ///
+  /// @param gctx the geometry contextfor this extraction call
+  /// @param nState is the current navigation state
+  /// @param indices is an ignored index vector
+  ///
+  /// @return a vector of raw Surface pointers
+  inline static const std::vector<const Surface*> extract(
+      [[maybe_unused]] const GeometryContext& gctx,
+      const NavigationState& nState,
+      [[maybe_unused]] const std::vector<size_t> indices = {}) {
+    if (nState.currentVolume == nullptr) {
+      throw std::runtime_error(
+          "AllSurfacesExtractor: no detector volume given.");
+    }
+    return nState.currentVolume->surfaces();
+  }
+};
+
+/// Helper extractors: indexed surfaces
+struct IndexedSurfacesExtractor {
```

Comment:
Could this and the other `struct`s just be free functions?

Reply:
I don't think I get it. It's pushed into a Delegate, isn't it? So this specific extractor (and the other structs) do not in fact have any members. I think in that case, I'd prefer these specific functions to just be free, even if others might have members and have to be structs.

Reply:
The point is that this implementation allows for both implementations of an extractor: one with a non-member structure where you have a `static` access method, another one where you extract from a member directly. 
As said, I used both implementations, because they might be useful.

Let me give two examples:
- An extractor of a volume from the detector based on a e.g. a grid or a BoundingBox hierarchy or whatsoever may use an extractor that has access to all volumes in the detector
- An extractor of e.g. a boundary surface of a container-type volume that is built from three volumes (before the entire detector is built and the final list of volumes known) may just keep memory of the three volumes it needs in a private member and thus needs to be a non-static member function

In `detray` we do not allow this second type of access, but that brings the consequence that every volume, when constructed needs to be provided by the detector in order to guarantee fixed indices at construction time of an impartially built detector, such as e.g.

```c++
auto& volume = detector.new_volume("DiamondSensorDetector");
```

I thought this would be too restrictive for the `Acts` core code.

Reply:
But isn't the ability to have members completely independent from this specific struct?